### PR TITLE
SCP-1986: Observable, accumulating state

### PR DESCRIPTION
**Pull request by j-mueller**
* Add an accumulating state `w` to contracts
  * (this is just a `Writer w`)
* After every update the we write out the new state `w` so that it can be observed from the outside
* How many type parameters is too many?
  * They all have a reason to be there
  * When we finally get around to looking at the API UX we could define a simplified type `Contract' = Contract ()` or whatever.
* @nau : You can use this to build the Marlowe companion contract

Pre-submit checklist:
- Branch
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [x] Relevant tickets are mentioned in commit messages
- PR
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [x] Reviewer requested
- If you updated any cabal files or added Haskell packages:
    - [ ] `nix-shell shell.nix --run updateMaterialized` to update the materialized Nix files
    - [ ] Update `hie-*.yaml` files if needed
- If you changed any Haskell files:
    - [x] `nix-shell shell.nix --run fix-stylish-haskell` to fix any formatting issues
- If you changed any Purescript files:
    - [ ] `nix-shell shell.nix --run fix-purty` to fix any formatting issues

Pre-merge checklist:
- [ ] Someone approved it
- [ ] Commits have useful messages
- [ ] Review clarifications made it into the code
- [ ] History is moderately tidy; or going to squash-merge

### CODE REVIEW

**Comment by j-mueller** [comment]
The main change is adding `Writer w` in `Language.Plutus.Contract.Types`. I also changed the auction contract to use the new `Writer`.

**Comment by michaelpj** [comment]
Dumb question: why do we need this? Can't a contract just expose a `queryState` endpoint and manage it itself if it needs thsi?

**Comment by j-mueller** [comment]
Endpoints don't return values, it doesn't fit into the model. They're events that you wait for, not function calls. I think about it like this:

* An endpoint `e` is like a `STM e` (returning when the endpoint has been called). We also have things like `awaitSlot :: Slot -> STM Slot` (returning when the target slot has been reached). 
  * So we can wait for changes on any of the variables we can observe. And the `select` primitive behaves exactly like `(<|>)` for `STM`.
  * This is also how the PAB is going to run contract instances without polling (https://github.com/input-output-hk/plutus/pull/2801)
* But then it's clear why we can't return a value to an endpoint call: Because there is no way to write back the value.
* With the `Writer w`, we have a way to write back the value
* To the outside the contract instance just looks like a `TVar w` - updating whenever the contract's state changes

So a `Contract` is just a constrainted version `STM` with additional effects for querying the chain index, building transactions, etc.

**Comment by michaelpj** [comment]
Ah, we have no way of sending anything back to the user at the moment? I thought we'd added that.

The `Writer w` approach lets us write a single kind of value, but I think we'll probably quickly hit the limits if we want to report different kinds of value for different endpoints or something.

I don't know how to make returning values to the user fit better, though.

**Comment by j-mueller** [comment]
I'm working on a better explanation of how it fits together, I'll ping you when it's done.

**Comment by j-mueller** [comment]
see https://github.com/input-output-hk/plutus/pull/2810

**Comment by j-mueller** [comment]
@michaelpj : In addition to the general documentation (#2810), this is what want to do specifically for Marlowe. 

* For each emulated wallet we have a "companion contract" that watches the wallet address
* When a new Marlowe role token is deposited at the wallet address, the companion contract extracts the address of the token's Marlowe contract instance from the token and starts watching it. Whenever the state of the Marlowe contract instance is changed by a transaction, the companion contract records the `Input` that was applied and the slot number it happened in.
* The writer type `w` of the companion contract is something like `Map MarloweRoleToken MarloweInstanceData`.
* When the user logs in, they can see their role tokens and the state of the Marlowe contracts they are participating in. Then they can start a separate contract instance for this Marlowe contract (if they wish) and engage with it.

That way we can minimise the amount of data that needs to be passed between participants, and we ensure that users can see the full history of a Marlowe contract even if they weren't logged in when the contract started. 

Generally the observable state is useful for things like currency hashes, tokens, account balances, etc. And another idea that we applied here is that the individual contract instances that are running inside the PAB shouldn't try to do too much. We can split up the work over several instances and let the client/frontend do the coordination.

**Comment by michaelpj** [comment]
Interesting. A view-only application that never makes transactions! I guess that makes sense.

I guess it's just the informal connection to what `w` means that makes me feel like it's a bit arbitrary. Like, we have all this carefully typed stuff for the endpoints, and then the response to them can just be emitting a random `w`, or several.

**Comment by j-mueller** [comment]
The reason why the endpoints are a much more complex type is to make it possible to signal to the outside world what kind of input is acceptable at the moment. But generally the `Contract` type now reads inputs of type `Event s` (= endpoints + blockchain stuff) and writes outputs of type `w`, with some effects for the chain index, building transactions, etc. I think that's a sensible model.

**Comment by michaelpj** [comment]
Yeah, it seems reasonble.

File: plutus-contract/src/Language/Plutus/Contract/Types.hs [hunk 1]
```diff
@@ -294,21 +305,24 @@ mkResult (((initialRes, cpKey), cpStore), newLogs) =
...
-  forall e i a.
-  CheckpointKey
+  forall w e i a.
+  Monoid w
+  => CheckpointKey
   -> CheckpointStore
-  -> Eff (SuspendedContractEffects e i) a
-  -> (((Either e a, CheckpointKey), CheckpointStore), Seq (LogMessage Value))
+  -> Eff (SuspendedContractEffects w e i) a
+  -> ((((Either e a, CheckpointKey), CheckpointStore), w), Seq (LogMessage Value)) -- FIXME
```
Comment (michaelpj) [inline plutus-contract/src/Language/Plutus/Contract/Types.hs:320]: fixme?

Comment (j-mueller) [reply plutus-contract/src/Language/Plutus/Contract/Types.hs:320]: Ah yeah I wanted to make that nested tuple a bit nicer.

Comment (j-mueller) [reply plutus-contract/src/Language/Plutus/Contract/Types.hs:320]: Will fix it.

**Review by michaelpj — APPROVED** [review]

**Pull request merged**